File: integer_square_root.c
#include <stdio.h>
#include <stdlib.h>

int i_sqrt(int num) {
    if (num < 2)
        return num;
    int sqrt = 0;
    int temp_sqrt = 0;
    int p = 0;
    int prev_p = 0;
    int temp_square = 0;
    int sum = 0;
    while (temp_sqrt * temp_sqrt <= num) {
        sqrt = temp_sqrt;
	p = 1;
	prev_p = 0;
        sum = temp_sqrt + p; 
	while ((sum * sum) <= num) {
	    prev_p = p;
	    p <<= 1;
            sum = temp_sqrt + p; 
	}
        if (prev_p == 0)
            break;
	temp_sqrt += prev_p;
    }
    return sqrt;
}

int main(int argc, char** argv) {
    if (argc < 2) {
        printf("The usage is:\n\t ./integer_square_root <integer>\n");
        return 1;
    }
    int num = atoi(argv[1]);
    int sqrt = i_sqrt(num);

    printf("Integer square root of %d is %d\n", num, sqrt); 
}

File: IntegerSquareRoot.java
public class IntegerSquareRoot {
	// O(log 2^31) = O( 31 log 2) = O(1)
	public int sqrt(int n) {
		if (n < 0)
			return -1;
		if (n < 2)
			return n;
		if (n < 4)
			return 1;
		int low = 1;
		int high = (n >> 1);
		int mid = 0;
		int sqr = 0;
		while (low <= high) {
			mid = (low + high) >> 1;
			sqr = mid * mid;
			if (sqr == n)
				return mid;
			else if (sqr < n) {
				if (low == mid)
					return mid;
				low = mid;
			} else
				high = mid;
		}
		return -1;
	}

	public static void main(String[] args) {
		IntegerSquareRoot isr = new IntegerSquareRoot();
		System.out.println(isr.sqrt(Integer.parseInt(args[0])));
	}
}
